### PR TITLE
[Dashboard] Start the new dashboard

### PR DESCRIPTION
**Pull request by fyrestone**
This PR contains the following features:
- set the `RAY_USE_NEW_DASHBOARD` environment var before `ray.init(include_dashboard=True)` will stat the new dashboard.
- raylet fork new dashboard agent if `RAY_USE_NEW_DASHBOARD` environment var exists.
- New dashboard head starts a grpc server (all head modules can extend the grpc server).
- New dashboard agent starts a http server (all agent modules can extend the http server).
- Add basic tests for new dashboard.
- `ClassMethodRouteTable` for head and agent http routes.
- Port features from reporter.py to reporter module.
- Some other improvements.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: dashboard/agent.py
Comment (rkooo567) [inline dashboard/agent.py:91]: Why not just sharing one?

File: dashboard/agent.py
Comment (rkooo567) [inline dashboard/agent.py:99]:
```suggestion
                "Dashboard agent suicides, "
```

File: dashboard/agent.py
Comment (rkooo567) [inline dashboard/agent.py:117]: Previously, we disabled cors when it was dev mode because frontend usually wants to start a react server. We aren't doing that anymore?

File: dashboard/agent.py
Comment (rkooo567) [inline dashboard/agent.py:147]: We should remove this from Redis when agents are shutdown. One of the most common workflow of Ray is using autoscaling clusters (and we will invest lots of resources on this), and this means agents will be killed occasionally from the cluster.

File: dashboard/head.py
Comment (rkooo567) [inline dashboard/head.py:140]:
```suggestion
                "Dashboard head suicides, "
```

File: dashboard/head.py
Comment (rkooo567) [inline dashboard/head.py:142]: When the head is killed, All the agents will still be alive?

File: dashboard/modules/reporter/reporter_head.py
Comment (rkooo567) [inline dashboard/modules/reporter/reporter_head.py:33]: Any case that index error can happen here?

File: python/setup.py
Comment (rkooo567) [inline python/setup.py:136]: No need for comment here.

File: python/setup.py
Comment (rkooo567) [inline python/setup.py:412]: Please use `{}` (for the consistent style)

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:252]: What is this?

File: src/ray/raylet/main.cc
Comment (rkooo567) [inline src/ray/raylet/main.cc:190]: Can you make it DEBUG

File: src/ray/raylet/agent_manager.cc
Comment (rkooo567) [inline src/ray/raylet/agent_manager.cc:86]: We should kill this thread in this case?

**Review by rkooo567 — COMMENTED** [review]
LGTM in general. I have several questions though. 

1. It is really lacking tests. We need A LOT of tests to make sure it works properly. What's you guys test plan?
2. What's the reason why each agents run http servers? It seems to be unncessary.
3. Also, what's the reason why the head dashboard has grpc endpoints?

File: dashboard/agent.py
Comment (fyrestone) [reply dashboard/agent.py:91]: The `aioredis_client` is shared by all the modules.

File: dashboard/agent.py
Comment (fyrestone) [reply dashboard/agent.py:147]: The agent will be restarted after killed, and the new value will be set to Redis. So, can we leave this from Redis when agents are shutdown?

File: dashboard/agent.py
Comment (fyrestone) [reply dashboard/agent.py:117]: Currently, the `CORS` is for the new `log` module. Each agent starts a http server for log file access. Allow new dashboard frontend to access log file as static resource directly.

File: dashboard/head.py
Comment (fyrestone) [reply dashboard/head.py:142]: Yes, agents are alive with raylet.

File: python/setup.py
Comment (fyrestone) [reply python/setup.py:136]: OK. I will remove the comments.

File: python/setup.py
Comment (fyrestone) [reply python/setup.py:412]: OK.

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (fyrestone) [reply src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:252]: `execute_after` returns a timer instance, so I add a `RAY_UNUSED` here.

File: src/ray/raylet/agent_manager.cc
Comment (fyrestone) [reply src/ray/raylet/agent_manager.cc:86]: The `monitor_thread` has been detached. So, the thread will be exit after this lambda returns.

File: src/ray/raylet/main.cc
Comment (fyrestone) [reply src/ray/raylet/main.cc:190]: OK

File: dashboard/modules/reporter/reporter_head.py
Comment (fyrestone) [reply dashboard/modules/reporter/reporter_head.py:33]: Currently, the ports are [http_port, grpc_port]. So, It can't raise an IndexError here.

**Comment by fyrestone** [comment]
> LGTM in general. I have several questions though.
> 
> 1. It is really lacking tests. We need A LOT of tests to make sure it works properly. What's you guys test plan?
> 2. What's the reason why each agents run http servers? It seems to be unncessary.
> 3. Also, what's the reason why the head dashboard has grpc endpoints?

1. Currently, I have added a basic test for the dashboard and agent, but the internal states can only be exposed by modules. So, I will add a `test` module which provides a `dump` API for more tests. BTW, The `reporter` module is not guaranteed OK in this PR. I think A LOT tests should be added to each modules.
2. We start a log server on each agent and it's easy to extend more features on agents.
3. Some of the data will be `pushed` to dashboard in the future, it will be more effective than `pull` the data periodically. The `pubsub` depends too much on the store implementation `Redis`.

**Review by mfitton — COMMENTED** [review]
It looks good! I had a couple minor comments, but overall it makes sense. I'm checking the code out locally now and making some last checks, but I'm willing to give my approval shortly. It's worth getting a review from someone else who has better C++ knowledge than I for parts of the code.

File: dashboard/head.py
Comment (rkooo567) [reply dashboard/head.py:142]: Who's in charge of restarting the head dashboard in this case?

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:252]: I mean is this relevant to this PR?

File: src/ray/raylet/agent_manager.cc
Comment (rkooo567) [reply src/ray/raylet/agent_manager.cc:86]: Ah, that makes sense. I was confused.

File: dashboard/agent.py
Comment (fyrestone) [reply dashboard/agent.py:117]: `dashboard` has the same access level with `agent`. We can access all routes of `dashboard`, why disallow access to all routes of `agent`?

The `log`http server on each `agent` can reduce the load of the `dashboard`.

File: dashboard/agent.py
Comment (fyrestone) [reply dashboard/agent.py:147]: No, If a node is removed from the cluster, the entry for the `agent` will not be cleaned up. `Dashboard` use the `node ip` to access agent entry, so the dead agent entry is not a problem.  If we really need to cleanup the dead agent entry, we can:
- cleanup the agent entry by GCS. (node can't ensure the cleanup)
- use TTL for the agent entry, agent will update the TTL periodically.

File: dashboard/head.py
Comment (fyrestone) [reply dashboard/head.py:140]: This version is better. I will change the text.

File: dashboard/head.py
Comment (fyrestone) [reply dashboard/head.py:142]: The one who restart the `GCS` has the responsibility to restart the `dashboard`. Maybe it's better to fork `dashboard` by `GCS`?

File: src/ray/common/ray_config_def.h
Comment (fyrestone) [reply src/ray/common/ray_config_def.h:337]: Thanks, I will clarify the comments.

File: src/ray/protobuf/BUILD
Comment (fyrestone) [reply src/ray/protobuf/BUILD:86]: `Dashboard` use a Python grpc client to access data from GCS, so this is necessary.

File: src/ray/raylet/agent_manager.cc
Comment (fyrestone) [reply src/ray/raylet/agent_manager.cc:40]: Thanks.

File: dashboard/modules/reporter/reporter_head.py
Comment (fyrestone) [reply dashboard/modules/reporter/reporter_head.py:81]: Thanks, I will rename the variables.

File: src/ray/raylet/agent_manager.cc
Comment (qicosmos) [inline src/ray/raylet/agent_manager.cc:63]: If failed to start a process, we need retry, so add retry logic here is better.
Another question is always retry or retry few times?

File: src/ray/util/process.cc
Comment (mehrdadn) [reply src/ray/util/process.cc:398]: Actually Boost.Process is what we tried before this. It turned out that Boost.Process was quite unstable and had bugs (and slowed down compilation dramatically too), and after struggling with it for over 1 month, we had to scrap it and roll our own solution, which works fine. Please don't revert back to it; it is very likely to cause problems again.

@fyrestone Regarding the `const`, this shouldn't be `const`. What was the reason for adding `const` here?

File: src/ray/raylet/agent_manager.cc
Comment (fyrestone) [reply src/ray/raylet/agent_manager.cc:63]: Currently, raylet always retry to start the agent. Yes, It's better to add the retry logic here, thanks.

File: src/ray/util/asio_util.h
Comment (fyrestone) [reply src/ray/util/asio_util.h:25]: Good suggestion.

File: src/ray/util/process.cc
Comment (qicosmos) [reply src/ray/util/process.cc:398]: > Actually Boost.Process is what we tried before this. It turned out that Boost.Process was quite unstable and had bugs (and slowed down compilation dramatically too), and after struggling with it for over 1 month, we had to scrap it and roll our own solution, which works fine. Please don't revert back to it; it is very likely to cause problems again.
> 
> @fyrestone Regarding the `const`, this shouldn't be `const`. What was the reason for adding `const` here?

Got it. Process could add some useful interface: WaitFor, WaitUntil.
Another suggestion, Process should forbidden copy construct and copy assign function, just move only will be better.
Hope you could add unit tests for Process, especially the exception tests.

Comment (mehrdadn) [reply src/ray/util/process.cc]: These are not trivial changes to make correctly. I don't think we should alter this specific class without compelling reasons.

The entire design of the API here was honestly quite deliberate. For example, I specifically allowed copying because the initial code used `pid_t` everywhere and this ensured semantic equivalence and minimized the changes of bugs being introduced. Requiring unique ownership would require a significant refactor of the rest of the codebase to accommodate it. If someone feels compelled to do it and can do it without introducing bugs, I suppose they can go ahead, but it is a significant effort and in the end there seems to be very little to gain.

Note that the primary intention for this class is to provide an easy-to-use interface that works correctly across all operating systems. While there _are_ certain features missing that could be added and potentially come in handy, the limited interface here was primarily designed to satisfy our actual needs without adding unnecessary maintenance burden. Process management is possibly the biggest source of pain when porting code across operating systems, and every change would have to be shown to work correctly on all 3 platforms with uniform semantics.

Overall, this is one of the few classes classes where it's not a good idea to change it just for the sake of elegance. If there is a functionality we actually *need*, then perhaps that warrants adding it. But as long as it is working correctly and other solutions can be reasonably implemented on top, it should probably be left alone.

Comment (fyrestone) [reply src/ray/util/process.cc]: > Actually Boost.Process is what we tried before this. It turned out that Boost.Process was quite unstable and had bugs (and slowed down compilation dramatically too), and after struggling with it for over 1 month, we had to scrap it and roll our own solution, which works fine. Please don't revert back to it; it is very likely to cause problems again.
> 
> @fyrestone Regarding the `const`, this shouldn't be `const`. What was the reason for adding `const` here?

I have removed const from Process::Kill().

File: src/ray/util/asio_util.h
Comment (fyrestone) [reply src/ray/util/asio_util.h:25]: But, It's better not to change the error check in this PR.

File: dashboard/agent.py
Comment (rkooo567) [reply dashboard/agent.py:147]: I think this is not a big deal for now if the only concern is having unused entries to GCS.

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Please make sure to pass all CIs. Also, please don't merge the PR until @mfitton approves!

**Review by mfitton — APPROVED** [review]

**Comment by mfitton** [comment]
Approved, but as Sang said, let's be careful not to break CI when merging.

**Pull request merged**
